incubator/tools-typescript/src/resolver.ts

```diff
@@ -0,0 +1,254 @@
+      return {
+        module,
+        suffix: suffixes ? suffixes[i] : "",
+        resolved: lookupModuleForFile(module, file),
+      };
+    });
+
+    // find the first unchecked or found value, if it is found return it
+    const first = entries.find((e) => !e.resolved || e.resolved.resolvedModule);
+    if (first && first.resolved?.resolvedModule) {
+      return first.resolved.resolvedModule;
+    }
+
+    // ensure the base entry has been resolved, this will be used to do file lookups
+    let result = entries.pop()!;
+    if (!result.resolved) {
+      const module = result.module;
+      result.resolved = resolver(module, file, redirected);
+      setModuleForFile(module, file, result.resolved);
+    }
+
+    // if entries still exist in the array see if files exist for those variants
+    const baseFile = result.resolved.resolvedModule?.resolvedFileName;
+    if (baseFile && entries.length > 0) {
+      const { base, ext } = parseSourceFileDetails(baseFile, true);
+
+      // if there is no extension this isn't a source file to try a different lookup for
+      if (ext) {
+        // now go through the non-base entries (.android, .native) to see if file variations exist
+        for (const entry of entries) {
+          const { module, suffix } = entry;
+          if (fs.existsSync(base + suffix + ext)) {
+            // if the file exists, have typescript try to find it
+            entry.resolved = resolver(module, file, redirected);
+          } else {
+            // if the file doesn't exist, bypass typescript and mark it as unfound
+            entry.resolved = { resolvedModule: undefined };
+          }
+          setModuleForFile(module, file, entry.resolved);
+          if (entry.resolved.resolvedModule) {
+            result = entry;
+            break;
+          }
+        }
+      }
+    }
+    return result.resolved?.resolvedModule;
+  });
+}
+
+/**
+ * Takes a file reference, either something like './folder/file' or './folder/file.js' and returns an array
+ * of modifications to try given the suffixes. For instance if the suffixes are ['.ios', '.native', '.js'] then
+ * - for './folder/file' returns './folder/file.ios', './folder/file.native', './folder/file.js'
+ * - for './folder/file.js' returns './folder/file.ios.js', './folder/file.native.js', './folder/file.js'
+ *
+ * @param fileRef the file reference to multiplex with suffixes, can be with or without extension
+ * @param suffixes suffixes to search
+ * @returns either just the original file ref or a search list in suffix order
+ */
+export function getFilesToSearch(
+  fileRef: string,
+  suffixes?: string[]
+): [string, ...string[]] {
+  if (suffixes && suffixes.length > 0) {
+    // split off the extension if it exists, then add the suffixes to try
+    const { base, ext = "" } = parseSourceFileDetails(fileRef, true);
+    return suffixes.map((s) => (s !== "." ? `${base}${s}${ext}` : fileRef)) as [
+      string,
+      ...string[],
+    ];
```

Comment:
Isn't this just `string[]`? If you fix the return type, we can drop the type casting:
```suggestion
    return suffixes.map((s) => (s !== "." ? `${base}${s}${ext}` : fileRef));
```

Reply:
Removed all of these since it still doesn't go through the map statement and it is just easier to add a single null check in the right place.